rust/lance-index/src/vector/ivf/shuffler.rs

```diff
@@ -148,25 +101,94 @@ pub async fn shuffle_dataset(
+            Schema::try_from(schema.as_ref())?,
+        )?;
+        unsafe {
+            shuffler.set_unsorted_buffers(&buffers);
+        }
 
-    let mut shuffler = IvfShuffler::try_new(
-        num_partitions,
-        num_sub_vectors,
-        None,
-        Schema::try_from(schema.as_ref())?,
-    )?;
+        shuffler
+    } else {
+        let mut shuffler = IvfShuffler::try_new(
+            num_partitions,
+            num_sub_vectors,
+            None,
+            Schema::try_from(schema.as_ref())?,
+        )?;
+
+        let precomputed_partitions = precomputed_partitions.map(Arc::new);
+        let stream = data
+            .zip(repeat_with(move || ivf.clone()))
+            .map(move |(b, ivf)| {
+                let col_ref = column.clone();
+
+                // If precomputed_partitions map is provided, use it
+                // for fast partitions.
+                let partition_map = precomputed_partitions
+                    .as_ref()
+                    .cloned()
+                    .unwrap_or(Arc::new(HashMap::new()));
+
+                tokio::task::spawn(async move {
+                    let batch = b?;
+
+                    let part_ids = if !partition_map.is_empty() {
+                        let row_ids = batch.column_by_name(ROW_ID).ok_or(Error::Index {
+                            message: "column does not exist".to_string(),
+                            location: location!(),
+                        })?;
+                        let part_ids = row_ids
+                            .as_primitive::<UInt64Type>()
+                            .values()
+                            .iter()
+                            .filter_map(|row_id| partition_map.get(row_id).copied())
+                            .collect::<Vec<_>>();
+                        Some(UInt32Array::from(part_ids))
+                    } else {
+                        None
+                    };
 
-    let start = std::time::Instant::now();
-    shuffler.write_unsorted_stream(stream).await?;
-    info!("wrote raw stream: {:?}", start.elapsed());
+                    ivf.partition_transform(&batch, col_ref.as_ref(), part_ids)
+                        .await
+                })
+            })
+            .buffer_unordered(num_cpus::get())
+            .map(|res| match res {
+                Ok(Ok(batch)) => Ok(batch),
+                Ok(Err(err)) => Err(Error::IO {
+                    message: err.to_string(),
+                    location: location!(),
+                }),
+                Err(err) => Err(Error::IO {
+                    message: err.to_string(),
+                    location: location!(),
+                }),
+            })
+            .boxed();
+
+        let stream = lance_core::io::RecordBatchStreamAdapter::new(schema.clone(), stream);
+
+        let start = std::time::Instant::now();
+        shuffler.write_unsorted_stream(stream).await?;
+        info!("wrote raw stream: {:?}", start.elapsed());
```

Comment:
Give some context in the log? For example: "IVFShuffler: write unsorted stream: {:?} seconds"?